File: windows/useragent.js
// #134: what User-Agent we tell websites we are.
//
// Electron's default UA is Chromium's own string with two extra tokens spliced
// in — the app name and the Electron version:
//
//   … (KHTML, like Gecko) WebForge/0.1.142 Chrome/132.0.0.0 Electron/34.5.8 Safari/537.36
//
// The engine really is Chromium 132, so sites that refuse us (Chase, and any
// bank with a "supported browsers" allowlist) are not detecting a missing
// capability — they are pattern-matching a brand they don't recognise and
// bailing out. Brave and Edge present as plain Chrome for the same reason.
//
// We derive the string by REMOVING those two tokens from whatever Electron
// generated, rather than hardcoding a UA of our own. That matters: a hardcoded
// string is correct on the day it's written and then silently rots into
// claiming an old Chrome every time we bump Electron, which is worse than not
// spoofing at all. Subtracting keeps `Chrome/NNN` truthful for free.
//
// Electron-free so the string surgery is testable without a display (see
// useragent.test.js).

// Both tokens are `Name/version`, always preceded by a space in Chromium's
// format. Taking the leading space with them is what keeps the result from
// containing a double space where the token used to be.
const DROP = [/ WebForge\/\S+/gi, / Electron\/\S+/gi];

/**
 * Strip WebForge's and Electron's tokens from a User-Agent string.
 * Anything else — platform, AppleWebKit, Chrome and Safari tokens — is left
 * exactly as Chromium wrote it.
 */
function cleanUserAgent(ua) {
  if (typeof ua !== 'string' || !ua) return '';
  let out = ua;
  for (const re of DROP) out = out.replace(re, '');
  // Defensive: if a future Electron formats the tokens differently and the
  // removal leaves ragged spacing, don't ship a UA with double spaces in it —
  // that alone looks synthetic to a sniffer.
  return out.replace(/\s{2,}/g, ' ').trim();
}

module.exports = { cleanUserAgent };

File: android/app/src/main/java/com/webforge/browser/UserAgent.kt
package com.webforge.browser

/**
 * #134: what User-Agent we tell websites we are.
 *
 * Sites with a "supported browsers" allowlist (Chase and most banks) refuse to
 * let you sign in when the UA names a browser they don't recognise — not
 * because anything is missing, but because they pattern-match a brand. The
 * Windows side has the same problem for a different reason (see
 * windows/useragent.js), so the two platforms fix it in the same spirit: take
 * whatever the engine generated and SUBTRACT the giveaway tokens, never
 * hardcode a UA of our own. A hardcoded string rots into claiming an old Chrome
 * as soon as the engine updates underneath it — and on Android the engine
 * updates through Play without us shipping anything at all, so a hardcoded UA
 * would rot silently and fast.
 *
 * Android System WebView's default looks like:
 *
 *   Mozilla/5.0 (Linux; Android 14; Pixel 7 Build/X; wv) AppleWebKit/537.36
 *     (KHTML, like Gecko) Version/4.0 Chrome/132.0.0.0 Mobile Safari/537.36
 *
 * Two tokens mark it as an embedded WebView rather than Chrome for Android:
 * the `; wv` inside the platform parentheses, and `Version/4.0` (an Android
 * Browser leftover that real Chrome does not send). Removing both leaves a
 * string identical to mobile Chrome's.
 */
object UserAgent {

    private val WV = Regex("""\s*;\s*wv""", RegexOption.IGNORE_CASE)
    private val VERSION = Regex("""\s*\bVersion/\S+""", RegexOption.IGNORE_CASE)
    private val GAPS = Regex("""\s{2,}""")

    /**
     * Strip the WebView giveaway tokens from a User-Agent string. Everything
     * else — platform, device, AppleWebKit, Chrome and Safari tokens — is left
     * exactly as the engine wrote it.
     */
    fun clean(ua: String?): String {
        if (ua.isNullOrEmpty()) return ""
        return ua
            .replace(WV, "")
            .replace(VERSION, "")
            // Defensive: never ship a UA with ragged spacing — that alone reads
            // as synthetic to a sniffer.
            .replace(GAPS, " ")
            .trim()
    }
}
